crates/stages/src/stages/execution.rs

```diff
@@ -313,6 +286,44 @@ impl<DB: Database> Stage<DB> for ExecutionStage {
         info!(target: "sync::stages::execution", stage_progress = end_block, done, "Sync iteration finished");
         Ok(ExecOutput { stage_progress: end_block, done })
     }
+}
+
+impl ExecutionStage {
+    /// Create new execution stage with specified config.
+    pub fn new(chain_spec: ChainSpec, commit_threshold: u64) -> Self {
+        Self { chain_spec, commit_threshold }
+    }
+}
+
+#[async_trait::async_trait]
+impl<DB: Database> Stage<DB> for ExecutionStage {
+    /// Return the id of the stage
+    fn id(&self) -> StageId {
+        EXECUTION
+    }
+
+    /// Execute the stage
+    async fn execute(
+        &mut self,
+        tx: &mut Transaction<'_, DB>,
+        input: ExecInput,
+    ) -> Result<ExecOutput, StageError> {
+        // For ethereum tests that has MAX gas that calls contract until max depth (1024 calls)
+        // revm can take more then default allocated stack space. For this case we are using
+        // local thread with increased stack size. After this task is done https://github.com/bluealloy/revm/issues/305
+        // we can see to set more accurate stack size or even optimize revm to move more data to
+        // heap.
+        std::thread::scope(|scope| {
+            let handle = std::thread::Builder::new()
+                .stack_size(50 * 1024 * 1024)
+                .spawn_scoped(scope, || {
+                    // execute and store output to results
+                    self.execute_inner(tx, input)
+                })
+                .expect("Expects that thread name is not null");
```

Comment:
This still blocks the task, right?

not sure how to make this async because the thread needs to borrow from Transaction I believe

Reply:
right... will fix that. I could use a simple thread there, don't need a local one.

Reply:
not sure if this is even easily possible atm because the execute functions needs to borrow from `Transaction` or at least is restricted by its lifetime.

but ideally we want to execute on a separate thread and then send the result through a oneshot channel back to the Stage

Reply:
having two separate `Transaction` would solve this problem, but unclear if applicable here.

since the Stage has exclusive ownership over the Transaction in `Stage::execute` we could think about turning this

```rust
async fn execute(&mut Transaction)
``

into

```rust
async fn execute(mut Transaction) -> Transaction
```

but because `thread` and even `task::spawn` require static lifetime we can't send anything that depends on Transaction to another thread/task inside `execute`.

Reply:
We can use scoped threads, no? https://doc.rust-lang.org/std/thread/fn.scope.html

Reply:
yes because a scope thread blocks the current thread.

Reply:
Yeah, I tried this and with lifetime it was not possible to  join threads outside of local thread.

Either way, having `mut Transaction` is worth considering to allow us more flexibility inside stages.

But definitely, the execution stage needs to be split between executing and applying  the changesets to the database.